File: windows/tests/ConnectorControl.Core.Tests/BackupManagerTests.cs
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests;

public class BackupManagerTests : IDisposable
{
    private const string Series = "claude_desktop_config";
    private readonly TempDir dir = new("backups");
    private readonly string source;
    private readonly BackupManager manager;

    public BackupManagerTests()
    {
        source = dir.File("claude_desktop_config.json");
        File.WriteAllText(source, "{\"mcpServers\": {}}");
        manager = new BackupManager(dir.File("backups"), keepCount: 3);
    }

    public void Dispose() => dir.Dispose();

    private static DateTime At(double unixSeconds) => DateTime.UnixEpoch.AddSeconds(unixSeconds);

    [Fact]
    public void BackUpCreatesTimestampedCopy()
    {
        var made = manager.BackUp(source, Series, At(1_752_600_000));
        Assert.NotNull(made);
        Assert.StartsWith("claude_desktop_config.", Path.GetFileName(made), StringComparison.Ordinal);
        Assert.EndsWith(".json", made, StringComparison.Ordinal);
        Assert.Equal(File.ReadAllBytes(source), File.ReadAllBytes(made));
    }

    [Fact]
    public void BackUpSkipsWhenIdenticalToNewest()
    {
        var first = manager.BackUp(source, Series, At(1_752_600_000));
        var second = manager.BackUp(source, Series, At(1_752_600_001));
        Assert.Equal(first, second);   // identical content returns the existing newest backup
        Assert.Single(manager.Backups(Series));
        File.WriteAllText(source, "changed");
        var third = manager.BackUp(source, Series, At(1_752_600_002));
        Assert.NotEqual(first, third);
        Assert.Equal(2, manager.Backups(Series).Count);
    }

    [Fact]
    public void BackUpDedupsOnlyAgainstNewest()
    {
        string[] contents = ["A", "B", "A"];
        for (int i = 0; i < contents.Length; i++)
        {
            File.WriteAllText(source, contents[i]);
            manager.BackUp(source, Series, At(1_752_600_000 + i));
        }
        Assert.Equal(3, manager.Backups(Series).Count);
    }

    [Fact]
    public void BackUpMissingSourceReturnsNull()
    {
        Assert.Null(manager.BackUp(dir.File("nope.json"), Series));
    }

    [Fact]
    public void RotationKeepsNewestKeepCount()
    {
        for (int i = 0; i < 5; i++)
        {
            File.WriteAllText(source, $"v{i}");
            manager.BackUp(source, Series, At(1_752_600_000 + i));
        }
        var kept = manager.Backups(Series);
        Assert.Equal(3, kept.Count);
        Assert.Equal("v4", File.ReadAllText(kept[0]));
        Assert.Equal("v2", File.ReadAllText(kept[2]));
    }

    /// <summary>Pruning is best effort: a stale backup that refuses deletion must not fail the write it trails.</summary>
    [Fact]
    public void RotationSkipsAStaleBackupThatCannotBeDeleted()
    {
        // A locked file blocks File.Delete on Windows; on Unix, unlink succeeds on an open
        // file, so this scenario cannot be forced there.
        if (!OperatingSystem.IsWindows())
        {
            Assert.Skip("Windows only");
            return;   // CA1416: the analyzer needs an explicit exit after the guard
        }
        File.WriteAllText(source, "v0");
        var oldest = manager.BackUp(source, Series, At(1_752_600_000));
        Assert.NotNull(oldest);
        using var locked = new FileStream(oldest!, FileMode.Open, FileAccess.Read, FileShare.Read);
        for (int i = 1; i <= 3; i++)
        {
            File.WriteAllText(source, $"v{i}");
            manager.BackUp(source, Series, At(1_752_600_000 + i));   // must not throw even though pruning oldest fails
        }
        Assert.True(File.Exists(oldest));                 // the locked backup survives the failed prune
        Assert.Equal(4, manager.Backups(Series).Count);   // one more than KeepCount, because the oldest refused deletion
    }

    [Fact]
    public void OriginalSnapshotWrittenOnceAndNeverPruned()
    {
        manager.EnsureOriginalSnapshot(source);
        File.WriteAllText(source, "changed");
        manager.EnsureOriginalSnapshot(source);   // second call: no-op
        var original = Path.Combine(manager.BackupsDir, "claude_desktop_config.original.json");
        Assert.Equal("{\"mcpServers\": {}}", File.ReadAllText(original));
        for (int i = 0; i < 5; i++)
        {
            File.WriteAllText(source, $"w{i}");
            manager.BackUp(source, Series, At(1_752_700_000 + i));
        }
        Assert.True(File.Exists(original));
        Assert.DoesNotContain(manager.Backups(Series), p => Path.GetFileName(p).Contains(".original.", StringComparison.Ordinal));
    }

    [Fact]
    public void SameMillisecondBackupsBothSucceed()
    {
        var now = At(1_752_600_000.123);
        File.WriteAllText(source, "v0");
        var first = manager.BackUp(source, Series, now);
        File.WriteAllText(source, "v1");
        var second = manager.BackUp(source, Series, now);
        Assert.NotNull(first);
        Assert.NotNull(second);
        Assert.NotEqual(first, second);
        Assert.True(File.Exists(first));
        Assert.True(File.Exists(second));
        Assert.Equal(2, manager.Backups(Series).Count);
    }

    [Fact]
    public void BackupsArePrivate()
    {
        if (!OperatingSystem.IsWindows())
        {
            Assert.Skip("Windows only");
            return;   // CA1416: the analyzer needs an explicit exit after the guard
        }
        var made = manager.BackUp(source, "mcps");
        Assert.NotNull(made);
        Assert.True(OwnerOnlyAcl.IsOwnerOnly(made));
    }

    [Fact]
    public void BackupsDirectoryAndOriginalSnapshotArePrivate()
    {
        if (!OperatingSystem.IsWindows())
        {
            Assert.Skip("Windows only");
            return;   // CA1416: the analyzer needs an explicit exit after the guard
        }
        manager.EnsureOriginalSnapshot(source);
        Assert.True(OwnerOnlyAcl.IsOwnerOnly(manager.BackupsDir));
        Assert.True(OwnerOnlyAcl.IsOwnerOnly(Path.Combine(manager.BackupsDir, "claude_desktop_config.original.json")));
    }

    /// <summary>A config symlinked into a dotfiles repo: the backup must be a snapshot of the bytes,
    /// not a copy of the link (which would read the live file forever, so no restore could ever go back).</summary>
    [Fact]
    public void BackupOfASymlinkedSourceIsARealSnapshot()
    {
        var real = dir.File(Path.Combine("dotfiles", "claude.json"));
        Directory.CreateDirectory(Path.GetDirectoryName(real)!);
        File.WriteAllText(real, "v1");
        var link = dir.File("linked_config.json");
        try
        {
            File.CreateSymbolicLink(link, real);
        }
        catch (Exception ex) when (ex is UnauthorizedAccessException or IOException)
        {
            // Creating a symlink needs Developer Mode or elevation; GitHub's hosted Windows runners are elevated, so under FailSkips=true (ci.runsettings) this skip would surface as a failure if that ever changed.
            Assert.Skip("symlink creation is not permitted in this environment");
            return;
        }
        var made = manager.BackUp(link, Series);
        Assert.NotNull(made);
        Assert.Null(new FileInfo(made!).LinkTarget);   // the backup is a regular file, not a link
        File.WriteAllText(real, "v2");
        Assert.Equal("v1", File.ReadAllText(made!));   // the snapshot does not follow the live file
    }

    [Fact]
    public void SeriesAreIndependent()
    {
        manager.BackUp(source, Series);
        manager.BackUp(source, "mcps");
        Assert.Single(manager.Backups(Series));
        Assert.Single(manager.Backups("mcps"));
    }

    [Fact]
    public void BackupsOfMissingDirIsEmpty()
    {
        Assert.Empty(new BackupManager(dir.File("never")).Backups(Series));
    }

    [Fact]
    public void AFreshBackupsTreeIsOwnerOnlyFromTheParentDown()
    {
        // First run with an empty Claude config: the backups dir is the first thing the app creates,
        // so its parent (the default store dir) must be private from the create call too.
        var backups = dir.File(Path.Combine("Connector Control", "backups"));
        var fresh = new BackupManager(backups, keepCount: 3);
        var made = fresh.BackUp(source, Series, At(1_752_600_000));
        Assert.NotNull(made);
        Assert.Equal(File.ReadAllBytes(source), File.ReadAllBytes(made));
        if (OperatingSystem.IsWindows())
        {
            Assert.True(OwnerOnlyAcl.IsOwnerOnly(Path.GetDirectoryName(backups)!), "the parent this call created is private");
            Assert.True(OwnerOnlyAcl.IsOwnerOnly(backups));
            Assert.True(OwnerOnlyAcl.IsOwnerOnly(made), "the copy is private from its create call, not a fix-up");
        }
    }
}

File: Tests/ConnectorControlCoreTests/BackupManagerTests.swift
import XCTest
@testable import ConnectorControlCore

final class BackupManagerTests: XCTestCase {
    var dir: URL!
    var source: URL!
    var manager: BackupManager!

    override func setUpWithError() throws {
        dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("backups-\(UUID().uuidString)")
        source = dir.appendingPathComponent("claude_desktop_config.json")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        try Data(#"{"mcpServers": {}}"#.utf8).write(to: source)
        manager = BackupManager(backupsDir: dir.appendingPathComponent("backups"), keepCount: 3)
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    func testBackupsDirectoryIsCreatedPrivate() throws {
        _ = try manager.backUp(fileAt: source, series: "claude_desktop_config")
        let mode = try XCTUnwrap(FileManager.default
            .attributesOfItem(atPath: manager.backupsDir.path)[.posixPermissions] as? Int)
        XCTAssertEqual(mode, 0o700)
    }

    func testBackUpCreatesTimestampedCopy() throws {
        let made = try XCTUnwrap(manager.backUp(
            fileAt: source, series: "claude_desktop_config",
            now: Date(timeIntervalSince1970: 1_752_600_000)))
        XCTAssertTrue(made.lastPathComponent.hasPrefix("claude_desktop_config."))
        XCTAssertTrue(made.lastPathComponent.hasSuffix(".json"))
        XCTAssertEqual(try Data(contentsOf: made), try Data(contentsOf: source))
    }

    func testBackUpSkipsWhenIdenticalToNewest() throws {
        let first = try XCTUnwrap(manager.backUp(
            fileAt: source, series: "claude_desktop_config",
            now: Date(timeIntervalSince1970: 1_752_600_000)))
        let second = try manager.backUp(
            fileAt: source, series: "claude_desktop_config",
            now: Date(timeIntervalSince1970: 1_752_600_001))
        // resolvingSymlinksInPath: the directory listing returns /private/var
        // URLs while the write path builds /var ones — same file.
        XCTAssertEqual(second?.resolvingSymlinksInPath(), first.resolvingSymlinksInPath(),
                       "identical content returns the existing newest backup")
        XCTAssertEqual(try manager.backups(series: "claude_desktop_config").count, 1)
        try Data("changed".utf8).write(to: source)
        let third = try XCTUnwrap(manager.backUp(
            fileAt: source, series: "claude_desktop_config",
            now: Date(timeIntervalSince1970: 1_752_600_002)))
        XCTAssertNotEqual(third, first)
        XCTAssertEqual(try manager.backups(series: "claude_desktop_config").count, 2)
    }

    func testBackUpDedupsOnlyAgainstNewest() throws {
        // A → B → back to A: the return to A still records — dedup compares
        // against the newest snapshot only, not the whole history.
        for (i, content) in ["A", "B", "A"].enumerated() {
            try Data(content.utf8).write(to: source)
            try manager.backUp(fileAt: source, series: "claude_desktop_config",
                               now: Date(timeIntervalSince1970: Double(1_752_600_000 + i)))
        }
        XCTAssertEqual(try manager.backups(series: "claude_desktop_config").count, 3)
    }

    func testBackUpMissingSourceReturnsNil() throws {
        let missing = dir.appendingPathComponent("nope.json")
        XCTAssertNil(try manager.backUp(fileAt: missing, series: "claude_desktop_config"))
    }

    func testRotationKeepsNewestKeepCount() throws {
        for i in 0..<5 {
            try Data("v\(i)".utf8).write(to: source)
            try manager.backUp(fileAt: source, series: "claude_desktop_config",
                               now: Date(timeIntervalSince1970: Double(1_752_600_000 + i)))
        }
        let kept = try manager.backups(series: "claude_desktop_config")
        XCTAssertEqual(kept.count, 3)
        XCTAssertEqual(try String(contentsOf: kept[0], encoding: .utf8), "v4")
        XCTAssertEqual(try String(contentsOf: kept[2], encoding: .utf8), "v2")
    }

    func testOriginalSnapshotWrittenOnceAndNeverPruned() throws {
        try manager.ensureOriginalSnapshot(of: source)
        try Data("changed".utf8).write(to: source)
        try manager.ensureOriginalSnapshot(of: source)  // second call: no-op
        let original = manager.backupsDir
            .appendingPathComponent("claude_desktop_config.original.json")
        XCTAssertEqual(try String(contentsOf: original, encoding: .utf8),
                       #"{"mcpServers": {}}"#)
        for i in 0..<5 {
            try manager.backUp(fileAt: source, series: "claude_desktop_config",
                               now: Date(timeIntervalSince1970: Double(1_752_700_000 + i)))
        }
        XCTAssertTrue(FileManager.default.fileExists(atPath: original.path))
        XCTAssertFalse(try manager.backups(series: "claude_desktop_config")
            .contains { $0.lastPathComponent.contains(".original.") })
    }

    func testSameMillisecondBackupsBothSucceed() throws {
        let now = Date(timeIntervalSince1970: 1_752_600_000.123)
        try Data("v0".utf8).write(to: source)
        let first = try XCTUnwrap(manager.backUp(
            fileAt: source, series: "claude_desktop_config", now: now))
        try Data("v1".utf8).write(to: source)
        let second = try XCTUnwrap(manager.backUp(
            fileAt: source, series: "claude_desktop_config", now: now))
        XCTAssertNotEqual(first, second)
        XCTAssertTrue(FileManager.default.fileExists(atPath: first.path))
        XCTAssertTrue(FileManager.default.fileExists(atPath: second.path))
        XCTAssertEqual(try manager.backups(series: "claude_desktop_config").count, 2)
    }

    func testBackupsArePrivate() throws {
        let made = try XCTUnwrap(manager.backUp(fileAt: source, series: "mcps"))
        let mode = try XCTUnwrap(FileManager.default
            .attributesOfItem(atPath: made.path)[.posixPermissions] as? Int)
        XCTAssertEqual(mode, 0o600)
    }

    /// Claude's own config is 0644; a copy would inherit that. The snapshot is
    /// written like every other private file instead.
    func testOriginalSnapshotOfAWorldReadableSourceIsPrivate() throws {
        try FileManager.default.setAttributes([.posixPermissions: 0o644], ofItemAtPath: source.path)
        try manager.ensureOriginalSnapshot(of: source)
        let original = manager.backupsDir.appendingPathComponent("claude_desktop_config.original.json")
        let mode = try XCTUnwrap(FileManager.default
            .attributesOfItem(atPath: original.path)[.posixPermissions] as? Int)
        XCTAssertEqual(mode, 0o600)
    }

    /// A config symlinked into a dotfiles repo: the backup must be a snapshot
    /// of the bytes, not a copy of the link (which would read the live file
    /// forever, so no restore could ever go back).
    func testBackupOfASymlinkedSourceIsARealSnapshot() throws {
        let fm = FileManager.default
        let real = dir.appendingPathComponent("dotfiles/claude.json")
        try fm.createDirectory(at: real.deletingLastPathComponent(), withIntermediateDirectories: true)
        try Data("v1".utf8).write(to: real)
        let link = dir.appendingPathComponent("linked_config.json")
        try fm.createSymbolicLink(at: link, withDestinationURL: real)
        let made = try XCTUnwrap(manager.backUp(fileAt: link, series: "claude_desktop_config"))
        XCTAssertNil(try? fm.destinationOfSymbolicLink(atPath: made.path), "the backup is a regular file")
        try Data("v2".utf8).write(to: real)
        XCTAssertEqual(try String(contentsOf: made, encoding: .utf8), "v1", "the snapshot does not follow the live file")
    }

    func testSeriesAreIndependent() throws {
        try manager.backUp(fileAt: source, series: "claude_desktop_config")
        try manager.backUp(fileAt: source, series: "mcps")
        XCTAssertEqual(try manager.backups(series: "claude_desktop_config").count, 1)
        XCTAssertEqual(try manager.backups(series: "mcps").count, 1)
    }
}
